src/track/steminfoimporter.cpp

```diff
@@ -0,0 +1,144 @@
+#include "track/steminfoimporter.h"
+
+#include <QFile>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QString>
+
+#include "util/logger.h"
+
+#define ATOM_TYPE(value)                                                   \
+    (uint32_t)(((uint8_t)(value)[0] << 24) | ((uint8_t)(value)[1] << 16) | \
+            ((uint8_t)(value)[2] << 8) | (uint8_t)(value)[3])
+
+namespace mixxx {
+
+namespace {
+
+const mixxx::Logger kLogger("StemInfoImporter");
+constexpr int kSupportedStemVersion = 1;
+const QStringList kStemFileExtensions = {".stem.mp4", ".stem.m4a"};
+
+const uint32_t kAtomHeaderSize = 8; // 4 bytes (unsigned integer) + 4 char
+const uint32_t kStemManifestAtomPath[] = {
+        ATOM_TYPE("moov"),
+        ATOM_TYPE("udta"),
+        ATOM_TYPE("stem"),
+        0 // This indicate end of the path
+};
+
+/// @brief Seek the reader the atom requested.
+/// @param reader The IODevice to search MP4 atom in
+/// @param path The list of atom to traverse in the tree, from top to bottom.
+/// Must be null terminated
+/// @param boxSize The size of the box currently under the cursor to focus the
+/// search in
+/// @param pathIdx The level of the tree currently search (index of path)
+/// @return the size of the box data if found, -1 otherwise
+uint32_t seekTillAtom(QIODevice& reader,
+        const uint32_t path[],
+        uint32_t boxSize = 0,
+        uint32_t pathIdx = 0) {
+    if (!path[pathIdx]) {
+        return boxSize;
+    }
+
+    uint32_t byteRead = 0;
+    char buffer[kAtomHeaderSize];
+    while (!reader.atEnd() || (boxSize && byteRead >= boxSize)) {
+        reader.read(buffer, kAtomHeaderSize);
+        byteRead += boxSize;
+        if (ATOM_TYPE(buffer + 4) == path[pathIdx]) {
+            return seekTillAtom(reader, path, ATOM_TYPE(buffer) - kAtomHeaderSize, pathIdx + 1);
```

Comment:
There's also the possibility of 64 bit box; in case `atomsize==1` there's an additional [8 bytes encoding the size](https://github.com/DigiDNA/ISOBMFF/blob/main/ISOBMFF/source/ContainerBox.cpp#L101).